cluster/juju/layers/kubernetes-worker/reactive/kubernetes_worker.py

```diff
@@ -121,6 +121,28 @@ def upgrade_charm():
     set_state('kubernetes-worker.restart-needed')
 
 
+@when('kubernetes-worker.remove-old-ingress')
+def remove_old_ingress():
+    kubectl_success('delete', 'rc', 'nginx-ingress-controller')
+
+    # these moved into a different namespace for 1.12
+    kubectl_success('delete', 'rc', 'default-http-backend')
+    kubectl_success('delete', 'svc', 'default-http-backend')
+    kubectl_success('delete', 'serviceaccount',
+                    'nginx-ingress-{}-serviceaccount'.format(
+                        hookenv.service_name()))
+    kubectl_success('delete', 'clusterrolebinding',
+                    'nginx-ingress-clusterrole-nisa-{}-binding'.format(
+                        hookenv.service_name()))
+    kubectl_success('delete', 'configmap',
+                    'nginx-load-balancer-{}-conf'.format(
+                        hookenv.service_name()))
+    kubectl_success('delete', 'ds',
+                    'nginx-ingress-{}-controller'.format(
+                        hookenv.service_name()))
+    remove_state('kubernetes-worker.remove-old-ingress')
```

Comment:
This is close to what I had in mind, but I think it still has the same problem: if any of these deletes fail, `kubectl_success` will return `False`, but the result isn't checked, it will fall through to the `remove_state` every time, and the deletes won't be retried.

I think you need to do two more things to make this behave well:
1. If any of these kubectl calls fails, do not clear the `kubernetes-worker.remove-old-ingress` flag. (You can check the result of `kubectl_success`, or use `kubectl` ([link](https://github.com/juju-solutions/kubernetes/blob/409c3b8940978f9d2fbc70a1c709e32a8bed6b3e/cluster/juju/layers/kubernetes-common/lib/charms/layer/kubernetes_common.py#L200)) and catch the CalledProcessError, whichever you prefer)
2. Add `--ignore-not-found` to the kubectl delete args so that a repeat attempt does not fail:
```
--ignore-not-found=false: Treat "resource not found" as a successful delete. Defaults to "true" when --all is
specified.
```

Reply:
Got it. For some reason, I had it in my head that kubectl_success would explode if it didn't succeed. I will update.